jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/DulmageMendelsohnDecomposition.java

```diff
@@ -0,0 +1,303 @@
+            while (it.hasNext()) {
+                subset2.add(it.next());
+            }
+        });
+        // Find the square subgraph
+        Set<V> subset3 = new HashSet<>();
+        subset3.addAll(partition1);
+        subset3.addAll(partition2);
+        subset3.removeAll(subset1);
+        subset3.removeAll(subset2);
+        if (fine) {
+            List<Set<V>> out = new ArrayList<>();
+            // Build a directed graph between edges of the matching in subset3
+            Graph<E, DefaultEdge> H = asDirectedEdgeGraph(matching, subset3);
+
+            // Perform strongly-connected-components on the graph
+            StrongConnectivityAlgorithm<E, DefaultEdge> sci
+                    = new KosarajuStrongConnectivityInspector<>(H);
+            // Divide into sets of vertices
+            for (Set<E> edgeSet : sci.stronglyConnectedSets()) {
+                Set<V> vertexSet = new HashSet<>();
+                edgeSet.stream().map((edge) -> {
+                    vertexSet.add(graph.getEdgeSource(edge));
+                    return edge;
+                }).forEachOrdered((edge) -> {
+                    vertexSet.add(graph.getEdgeTarget(edge));
+                });
+                out.add(vertexSet);
+            }
+            return new Decomposition<>(graph, subset1, subset2, out);
+        } else {
+            return new Decomposition<>(graph, subset1, subset2, Arrays.asList(subset3));
+        }
+    }
+
+    /**
+     * The output of a decomposition operation
+     *
+     * @param <V>
+     * @param <E>
+     */
+    public static class Decomposition<V, E> {
+
+        private final Graph<V, E> graph;
+        private final Set<V> subset1;
+        private final Set<V> subset2;
+        private final List<Set<V>> subset3;
+
+        Decomposition(Graph<V, E> graph, Set<V> subset1, Set<V> subset2, List<Set<V>> subset3) {
+            this.graph = graph;
+            this.subset1 = subset1;
+            this.subset2 = subset2;
+            this.subset3 = subset3;
+        }
+
+        /**
+         * Gets the subset dominated by partition1
+         *
+         * @return Set of vertices in the subset
+         */
+        public Set<V> getPartition1DominatedSet() {
+            return subset1;
+        }
+
+        /**
+         * Gets the subset dominated by partition2
+         *
+         * @return Set of vertices in the subset
+         */
+        public Set<V> getPartition2DominatedSet() {
+            return subset2;
+        }
+
+        /**
+         * Gets the remaining subset, or subsets in the fine decomposition
+         *
+         * @return List of Sets of vertices in the subsets
+         */
+        public List<Set<V>> getPerfectMatchedSets() {
+            return subset3;
+        }
+
+        /**
+         * Gets the subgraph dominated by partition1
+         *
+         * @return Subgraph
+         */
+        public Graph<V, E> getPartition1DominatedSubgraph() {
```

Comment:
I'm not sure what the purpose of these graph methods (getPartition1DominatedSubgraph, getPartition2DominatedSubgraph, getPerfectMatchedSubgraphs) is. I could not find them in the paper you are referencing. Are these methods just convenient for your application? If so, I would recommend deleting them as they are not really general purpose. Or do these graphs somehow relate to the over-constraint/well-constraint/under-constraint subgraphs mentioned in the paper. Either remove them if they are not general-purpose, or retain them and improve their documentation. Certainly add:
`Returns the subgraph induced by the vertices in the partition returned by {@link getPartition1DominatedSet}.`.

Reply:
The subgraphs could be created externally so I could remove those methods. I'll have a think about the description - for my use-case and the referenced paper it is analysis of equations and variables and hence over/under constrained has specific meaning, but I was trying to generalise to anything where there are 2 partitions.